File: libs/east-node/packages/east-node-cli/test-fixtures/crosslang-functions.mjs
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

/**
 * TypeScript-authored functions the python cross-import test imports (#628):
 * `east-node export-functions test-fixtures/crosslang-functions.mjs -o …`
 * writes their manifest; python links and runs them with no TypeScript at
 * run time. Pure functions only — no platform calls to provide.
 */

import { East, ArrayType, FloatType, IntegerType, StringType, StructType } from "@elaraai/east";

const Row = StructType({ qty: IntegerType, price: FloatType });

export const score = East.function([Row], FloatType, ($, r) => r.qty.toFloat().multiply(r.price));

export const total = East.function([ArrayType(Row)], FloatType, ($, rows) =>
    rows.map(($, r) => score(r)).reduce(($, acc, x) => acc.add(x), 0.0));

export const greet = East.function([StringType, IntegerType], StringType, ($, name, times) =>
    name.concat("!").repeat(times));

export const eastFunctions = { score, total, greet };

File: libs/east-py/packages/east-py/tests/fixtures/crosslang_functions.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""python-authored functions the TypeScript cross-import spec imports (#628).

Their manifest is checked in at ``libs/east/test/fixtures/py-functions.beast2``
(``EAST_UPDATE_FIXTURES=1 pytest tests/conformance/test_cross_import.py``
rewrites it; the test otherwise asserts the file still matches these
functions). Pure functions only — no platform calls to provide.
"""

from __future__ import annotations

from east import East
from east.types.types import ArrayType, FloatType, IntegerType, StringType, StructType

Row = StructType([("qty", IntegerType), ("price", FloatType)])

score = East.function([Row], FloatType, lambda b, r: r.qty.to_float() * r.price)

total = East.function(
    [ArrayType(Row)], FloatType,
    lambda b, rows: rows.map(lambda b, r: score(r)).reduce(lambda b, acc, x: acc + x, 0.0))

greet = East.function(
    [StringType, IntegerType], StringType,
    lambda b, name, times: name.concat("!").repeat(times))

east_functions = {"score": score, "total": total, "greet": greet}
